satellite/notifications/chore.go

```diff
@@ -0,0 +1,167 @@
+	defer mon.Task()(nil)(&err)
+	chore.Cycle.Close()
+	return nil
+}
+
+// report sends reputation reports to all reliable nodes.
+func (chore *Chore) report(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	sendReports := func(ctx context.Context, nodes []*overlay.NodeDossier) error {
+		for _, node := range nodes {
+			if err = ctx.Err(); err != nil {
+				return err
+			}
+
+			chore.sendReport(ctx, node)
+		}
+
+		return nil
+	}
+
+	return chore.iterate(ctx, sendReports)
+}
+
+// iterate iterates through batches of reliable nodes.
+func (chore *Chore) iterate(ctx context.Context, fn func(context.Context, []*overlay.NodeDossier) error) (err error) {
+	defer mon.Task()(&ctx)(nil)
+
+	var offset int64
+
+	// use default lookup limit
+	nodes, more, err := chore.overlay.PaginateReliable(ctx, offset, 0)
+	if err != nil {
+		return err
+	}
+
+	if err = fn(ctx, nodes); err != nil {
+		return err
+	}
+
+	for more {
+		if err = ctx.Err(); err != nil {
+			return err
+		}
+
+		nodes, more, err = chore.overlay.PaginateReliable(ctx, offset+int64(len(nodes)), 0)
+		if err != nil {
+			return err
+		}
+
+		if err = fn(ctx, nodes); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// sendReport sends report for provided node.
+func (chore *Chore) sendReport(ctx context.Context, node *overlay.NodeDossier) {
+	defer mon.Task()(&ctx)(nil)
+
+	logger := chore.log.Named(fmt.Sprintf("reporter %s", node.Id.String()))
+
+	client, err := NewClient(ctx, chore.dialer, node.Id, node.GetAddress().GetAddress())
+	if err != nil {
+		logger.Debug("failed to establish connection", zap.Error(err))
+		return
+	}
+
+	auditScore := calculateReputationScore(
+		node.Reputation.AuditReputationAlpha,
+		node.Reputation.AuditReputationBeta,
+	)
+	uptimeScore := calculateReputationScore(
+		node.Reputation.UptimeReputationAlpha,
+		node.Reputation.UptimeReputationBeta,
+	)
+
+	reportRequest := &pb.ReportRequest{
+		AuditTotalCount:       node.Reputation.AuditCount,
+		AuditSuccessCount:     node.Reputation.AuditSuccessCount,
+		AuditReputationAlpha:  node.Reputation.AuditReputationAlpha,
+		AuditReputationBeta:   node.Reputation.AuditReputationBeta,
+		AuditReputationScore:  auditScore,
+		UptimeTotalCount:      node.Reputation.UptimeCount,
+		UptimeSuccessCount:    node.Reputation.UptimeSuccessCount,
+		UptimeReputationAlpha: node.Reputation.UptimeReputationAlpha,
+		UptimeReputationBeta:  node.Reputation.UptimeReputationBeta,
+		UptimeReputationScore: uptimeScore,
+		LastContactSuccess:    node.Reputation.LastContactSuccess,
+		LastContactFailure:    node.Reputation.LastContactFailure,
+		ExitLoopCompletedAt:   node.ExitStatus.ExitLoopCompletedAt,
+		ExitInitiatedAt:       node.ExitStatus.ExitInitiatedAt,
+		ExitFinishedAt:        node.ExitStatus.ExitFinishedAt,
+		ExitSuccess:           node.ExitStatus.ExitSuccess,
+		Vetted:                !chore.overlay.IsNew(node),
+	}
+
+	if _, err = client.Report(ctx, reportRequest); err != nil {
```

Comment:
A slow node could end up blocking the whole thing.

Reply:
could make it async as well